### ISSUE TITLE
Change language_level for documentation examples to Python 3

### ISSUE THREAD

**Issue by da-woods** [cython/cython#5066]
It looks like not all of the documentation examples work right with language_level set to 3 (the default) - see #5065 as an example.

I don't think we want to change the test-suite as a whole right now, but the public-facing documentation would be better tested with the settings that people will actually be using.

**Comment by alexandrasp** [comment]
hello @da-woods, I am considering working on this bug now, is it ok? I am asking because maybe you could have any special suggestions since all issues with `help wanted` and `good first issues` are a bit old. :}

**Comment by da-woods** [comment]
I think this still needs doing even if old.

We just haven't been hugely good at keeping up with "help wanted" and "good first issue" tags.

In `runtest.py` there's an `examples_dir` - those need to be run with language_level=3 and any issues relating to that fixed.

**Comment by alexandrasp** [comment]
opened the PRs #6139 and #6140, but more are coming, the goal is to update everything from the examples dir.

**Comment by alexandrasp** [comment]
#6143 is supposed to solve all problems related to this issue, but while checking everything in `examples_dir` I stumbled into some edge cases.

Compiling [c_property.pyx](https://github.com/cython/cython/blob/master/docs/examples/userguide/extension_types/c_property.pyx#L7) I got: 
```warning: docs/examples/userguide/extension_types/c_property.pyx:7:4: complex already a builtin Cython type```

Should I change this example to something not defined in cython builtin?

-----------------------------------------

Compiling [casting_python.pyx](https://github.com/cython/cython/blob/master/docs/examples/userguide/language_basics/casting_python.pyx#L4) the output was: 
```warning: docs/examples/userguide/language_basics/casting_python.pyx:4:4: 'Py_intptr_t' redeclared```

 Do you think I should change this?

-----------------------------------------

From [memory_layout_2.pyx](https://github.com/cython/cython/blob/master/docs/examples/userguide/memoryviews/memory_layout_2.pyx): 

```bash
Error compiling Cython file:
    cdef int[::view.indirect, ::1, :] a
    cdef int[::view.indirect, :, ::1] b
    cdef int[::view.indirect_contiguous, ::1, :] c

    # INVALID
    cdef int[::view.contiguous, ::view.indirect, :] d
            ^
------------------------------------------------------------

docs/examples/userguide/memoryviews/memory_layout_2.pyx:11:12: Only dimension 2 may be contiguous and direct

Error compiling Cython file:
------------------------------------------------------------
...
    cdef int[::view.indirect, :, ::1] b
    cdef int[::view.indirect_contiguous, ::1, :] c

    # INVALID
    cdef int[::view.contiguous, ::view.indirect, :] d
    cdef int[::1, ::view.indirect, :] e
            ^
------------------------------------------------------------

docs/examples/userguide/memoryviews/memory_layout_2.pyx:12:12: Indirect dimension may not follow Fortran contiguous dimension
```

In this case, I assume everything is as expected because the errors are in the file. Is it correct?

------------------------------------------------------------

In [quickstart.pyx](https://github.com/cython/cython/blob/master/docs/examples/userguide/memoryviews/quickstart.pyx) and [memoryview_sum.pyx](https://github.com/cython/cython/blob/master/docs/examples/userguide/parallelism/memoryview_sum.pyx)

The compiler is sending a warning suggesting to use a `boundscheck(False)`. Do you think that is needed to update the code or just use the arg `-X boundscheck=False` is fine? 

-----------------------------------------------------------

In [condition_sum.pyx](https://github.com/cython/cython/blob/master/docs/examples/userguide/parallelism/condition_sum.pyx) we have a "special case". I am using `Cython version 3.0.10` and according to the [change log](https://cython.readthedocs.io/en/latest/src/changes.html) and I also checked the code, `prange` doesn't support the `use_threads_if` yet, but the code seems to be good.

-----------------------------------------------------------

Apart from the comments mentioned,  `cython -3 / cython -3 -+` commands are working as expected.

**Comment by da-woods** [comment]
> Compiling [c_property.pyx](https://github.com/cython/cython/blob/master/docs/examples/userguide/extension_types/c_property.pyx#L7) I got: `warning: docs/examples/userguide/extension_types/c_property.pyx:7:4: complex already a builtin Cython type`
> 
> Should I change this example to something not defined in cython builtin?

No - I think it's a good simple example.



> Compiling [casting_python.pyx](https://github.com/cython/cython/blob/master/docs/examples/userguide/language_basics/casting_python.pyx#L4) the output was: `warning: docs/examples/userguide/language_basics/casting_python.pyx:4:4: 'Py_intptr_t' redeclared`
> 
> Do you think I should change this?

Yes I think so - if you can just get rid of the typedef that'd be good (probably also applies to the pure Python example of the same thing)

> In this case, I assume everything is as expected because the errors are in the file. Is it correct?

Yes - that's expected

> In [quickstart.pyx](https://github.com/cython/cython/blob/master/docs/examples/userguide/memoryviews/quickstart.pyx) and [memoryview_sum.pyx](https://github.com/cython/cython/blob/master/docs/examples/userguide/parallelism/memoryview_sum.pyx)

Personally I'd leave those - the example still works without boundscheck and adding more things makes it harder to read quickly.

> In [condition_sum.pyx](https://github.com/cython/cython/blob/master/docs/examples/userguide/parallelism/condition_sum.pyx) we have a "special case". I am using `Cython version 3.0.10` and according to the [change log](https://cython.readthedocs.io/en/latest/src/changes.html) and I also checked the code, `prange` doesn't support the `use_threads_if` yet, but the code seems to be good.

Yes - the docs are for the master branch so include things not yet in a release (although you can also view the docs for the release branch). No need to change anything there.

**Comment by alexandrasp** [comment]
> Compiling [casting_python.pyx](https://github.com/cython/cython/blob/master/docs/examples/userguide/language_basics/casting_python.pyx#L4) the output was: `warning: docs/examples/userguide/language_basics/casting_python.pyx:4:4: 'Py_intptr_t' redeclared`
>  Do you think I should change this?
> 
> > Yes I think so - if you can just get rid of the typedef that'd be good (probably also applies to the pure Python example of the same thing)
>
Included in the PR.

**Comment by alexandrasp** [comment]
My final question regarding this issue is if I should keep the `-3` arg option as it is ? Is it important to keep the support to cy2?

**Comment by da-woods** [comment]
> My final question regarding this issue is if I should keep the `-3` arg option as it is ? Is it important to keep the support to cy2?

Not completely sure what you mean by that.

However, what I'd suggest is changing the behaviour in `runtest.py`:

https://github.com/cython/cython/blob/b2965bd476752da410ad5a62447185841da5aa8a/runtests.py#L2848-L2854

This block controls the testing of the examples directory. I think that should just be changed to use `language_level=3` and ignore the command-line argument.

**Comment by alexandrasp** [comment]
> > My final question regarding this issue is if I should keep the `-3` arg option as it is ? Is it important to keep the support to cy2?
> 
> Not completely sure what you mean by that.
> 
> However, what I'd suggest is changing the behaviour in `runtest.py`:
> 
> https://github.com/cython/cython/blob/b2965bd476752da410ad5a62447185841da5aa8a/runtests.py#L2848-L2854
> 
> This block controls the testing of the examples directory. I think that should just be changed to use `language_level=3` and ignore the command-line argument.

Thanks, you got it right. I got distracted by the option `-3` used by other parts of the code.

**Comment by da-woods** [comment]
Closed d17f7460c6558bf2790acfacddd99c9d52c25a8d
